pilot/pkg/networking/core/v1alpha3/dns_listener.go

```diff
@@ -0,0 +1,111 @@
+	dnstable "github.com/envoyproxy/go-control-plane/envoy/data/dns/v3"
+	dnsfilter "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/udp/dns_filter/v3alpha"
+	stringmatcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+)
+
+// This is a UDP listener is on port 15013 (TODO customize me via mesh config).
+// It has a DNS listener filter containing the cluster IPs of all services visible to the proxy (those that have one anyway).
+// The list of service cluster IPs will aid VMs, and multi-cluster setups to resolve services that may not exist in the local
+// cluster.
+//
+// TODO: Expand the logic to include necessary IPs for service entries that have NONE resolution with wildcard hosts
+// like *.example.com, or those that have service entries for TCP services with DNS resolution, without cluster IP.
+// In these cases, just allocate some dummy IP from
+// 127.255.0.0/16 subnet. But make sure that we are also building a TCP listener on this port to process this
+// traffic appropriately. These IPs need not be consistent across all proxies because DNS resolution is local to the pod/VM.
+// They need to be consistent between the IP configured in the DNS resolver here and the IP configured in the listeners
+// sent to this proxy. Once this system works properly, we should get rid of all k8s dns hacks
+func (configgen *ConfigGeneratorImpl) buildSidecarDNSListener(node *model.Proxy, push *model.PushContext) *listener.Listener {
+	_, localhost := getActualWildcardAndLocalHost(node)
+
+	address := util.BuildAddress(localhost, model.SidecarDNSListenerPort)
+	// Convert the address to a UDP address
+	address.GetSocketAddress().Protocol = core.SocketAddress_UDP
+
+	inlineDNSTable := configgen.buildInlineDNSTable(node, push)
+
+	dnsFilterConfig := &dnsfilter.DnsFilterConfig{
+		StatPrefix: "dns",
+		ServerConfig: &dnsfilter.DnsFilterConfig_ServerContextConfig{
+			ConfigSource: &dnsfilter.DnsFilterConfig_ServerContextConfig_InlineDnsTable{InlineDnsTable: inlineDNSTable},
+		},
+	}
+	dnsFilter := &listener.ListenerFilter{
+		Name: util.EnvoyDNSListenerFilterName,
+		ConfigType: &listener.ListenerFilter_TypedConfig{
+			TypedConfig: util.MessageToAny(dnsFilterConfig),
+		},
+	}
+
+	return &listener.Listener{
+		Name:             dnsListenerName,
+		Address:          address,
+		ListenerFilters:  []*listener.ListenerFilter{dnsFilter},
+		TrafficDirection: core.TrafficDirection_OUTBOUND,
+	}
+}
+
+func (configgen *ConfigGeneratorImpl) buildInlineDNSTable(node *model.Proxy, push *model.PushContext) *dnstable.DnsTable {
+
+	// build a virtual domain for each service visible to this sidecar
+	virtualDomains := make([]*dnstable.DnsTable_DnsVirtualDomain, 0)
+
+	knownSuffixes := []*stringmatcher.StringMatcher{
+		{
+			MatchPattern: &stringmatcher.StringMatcher_SafeRegex{
+				SafeRegex: &stringmatcher.RegexMatcher{
```

Comment:
is `prefix: ""` the same as `regex: .*`? might be a bit quicker

Reply:
empty prefix is not allowed.. 
```
Proto constraint validation failed (DnsFilterConfigValidationError.ServerConfig: ["embedded message failed validation"] | caused by ServerContextConfigValidationError.InlineDnsTable: ["embedded message failed validation"] | caused by DnsTableValidationError.KnownSuffixes[i]: ["embedded message failed validation"] | caused by StringMatcherValidationError.Prefix: ["value length must be at least " '\x01' " bytes"]): 
```